### ISSUE TITLE
Multi location support

### ISSUE THREAD

**Issue by jekkos** [jekkos/opensourcepos#150]
Currrently I'm working on a client request for multistore support for this project. The first iteration willl include adding a new table to the system in which pos locations can be defined

* pos location is considered as a browser accessing a central server, capable of doing sales, receivings, reporting or item management for specific stock locations
* employees will have a dropdown to choose a pos location to login to
* pos location will be saved in the employee's session
* grants will be specific to a pos location. In this case there'll be a dropdown in case there is more than one pos location to configure grants for a specific employee
* employees can only login to one pos location when they have at least one grant for that location
* stock locations have to be granted for every pos location as well. This will allow implementors to choose which pos location to sell/receive from 
* store config might have to be customized depending on the pos location. 

For this last part I wonder whether there's anyone having any suggestions as to which options (beside pos location name) might need to be customized in that case.

**Comment by jekkos** [comment]
Currently there's also a bug that prevents to do sales or receivings with invoicing numbers for multiple pos locations. This will need to be tackled as well. Issue can be found on sourceforge

https://sourceforge.net/p/opensourcepos/bugs/49/

**Comment by odiea** [comment]
Other than the Invoicing issue are there any other multi user issues. Say 2 cashiers and 3 people in the back room inputting Items. I have not tried this out yet to see how multi user works.

**Comment by jekkos** [comment]
I haven't done any real tests for these scenario's either, but I did some adaptations for another client some time ago, using it from three different browsers. He has reported the same problem as I have mentioned above here.

**Comment by daN4cat** [comment]
Ids should be generated at submission time and be unique.
Isn't that a caching issue with let say temp tables or similar?
It needs some deep analysis to check that part.

Also on a slightly different note but still about "multi" use, my wife always asked me to have a common login for let say sales assistant (maybe tagged to physical pos) and just select the seller name at transaction time selecting an icon (avatar) to make it easy with a click.

What's happening at the moment is that 1 girl logs in and everybody uses that log for the whole day if not week as they leave stuff run.

**Comment by odiea** [comment]
The other issue might be printer selection for multi user cashiers or Inputters. Say pos 1 has an Epson and pos 2 has a star printer. The folks in the back room printing labels all have different Epson or zebra printers. Maybe a multi congfig for each station and a cashier login log out without having to exit the program. All stuff to think about for the next generation of OSPOS

**Comment by daN4cat** [comment]
I was wondering about the model here, you really need to design on paper all the physical entities and then draw relationship and cardinality. Not sure the current database is organised in a scalable way, the code probably even less.

It's either "hack it up" for the specific need as 1st leg and refactor bit a bit, or refactor it more seriously and build more complex features.

Being pragmatic would be probably better the 1st approach even if not ideal because more grounds up  could be a massive work and a long lead.

**Comment by odiea** [comment]
Jekko has made a lot of improvements in this program since he has taken over. And thank you for your code improvements too. As a tester I cant imagine the hours you all have spent on this program making it better than the original I first started testing. It's come a long way.

**Comment by jekkos** [comment]
Yes in this case it will involve different cashiers using the pos from a separate locations. They'll have to login / logout every time someone starts or stops working. I agree that most of the time it's not practical to login/logout when you're all working from the same machine. But that's not really the usecase here.

I did some analysis on this on paper and up til now I found that what I describe above here be the most pragmatic solution. sales and receivings are the only real entities that can be bound to a pos_location. In this case one store can have mulitple cashiers selling from different stock locations.

That's why I think that modifying the grants would be a good solution as it will allow you to give access for an employee to specific modules at a specific location for a selection of stock locations at hand (as each stock location has it's own permission)

And indeed, the sale id generation issue can probably be solved by generating it at completion time, instead of storing in the session as it's done now. It's quite important to keep the sale_id as a single primary key, so assuming that all of the pos's are connected to the same database. If not then there will be other solutions that involve db partitioning or replication that can assure you to keep those id's unique.

The printing options were already refactord to be persisted in the browsers itself, as it doesn't make sense indeed to store this config on a server (most of the time you print directly from the wrokstation that has the browser installled)

**Comment by jekkos** [comment]
Second iteration will include adding cut z and cut x reports. I think this is a feature that has been requested earlier.

**Comment by daN4cat** [comment]
How far is a Z report is from the Takings we created?
The idea was about that but maybe you have more use cases there.

**Comment by daN4cat** [comment]
Just wondering, any impact on inventory table and suspended features?

**Comment by jekkos** [comment]
Ok I forgot that one, suspended sales will need to have those id's as well. Unless of course you want to unsuspend them at a different pos, but this isn't really a requirement right now for me.

The cut x and cut z should have some extra flow logic encapsulated into it. For example once a cashier opens or closes his session, he will need to make a report. This report is also required after certain periods of the day;

**Comment by daN4cat** [comment]
Do you even need to keep track of the petty cash at in/out times?
That's a common scenario in shops, no idea if that needs to be bound to a cashier.

**Comment by odiea** [comment]
I would say it would be necessary at the beginning of the day (add amount to a new file? to show what the opening cash was.) and the end of the day ( a report showing what the total cash in the register is Including opening cash amount) so cash sales come out properly

**Comment by odiea** [comment]
Would it be possible to setup Terminal ID,s. Cashier 1 terminal 1 Cashier 2 Terminal 2. Then the only Config that needs to be changed would be the printer selection locked to the terminal id.

**Comment by jekkos** [comment]
Yes indeed, I need opening and closing amounts and register amounts whenever a 'cut x' report is run. In this case I've added an ospos_pos_locations table (any suggestions on the name here?) that includes an id and name.

User can then select the pos location at login time (to which he has at least one grant). Next step is to extend grants settings in employees with a dropdown. Every module will become accessible per pos location. By default there's only one (as with stock locations) and so every pos location can have access to one or more stock locations.

Regarding the config, I think atm I will keep the config as global, except for the printer settings, which are already bound to the browser (and probably pos location)

**Comment by jekkos** [comment]
To make this work there'll has to be some 1-1 mapping between browser on a client and pos location. Any idea which criteria could be used to identify a client unquely on the server? (and thus assign it's pos_id)

**Comment by jekkos** [comment]
I think I'll solve that one by storing the encrypted pos_id in the browser's localstorage and by foreseeing a config page (only accessible by admin) to store this token locally.

This way a user can't really tamper with the mapping (as it should just be assigned once, during initial setup in the cashier's browser). This token has to be sent on login and then the pos can check whether this browser was assigned to the location at hand.

**Comment by odiea** [comment]
The way Traxia has it setup is they use terminal ids (Like the current custom fields the name can be aliased for easier consumer use) so the name and password will log you into the terminal. Once your in the terminal the user enters his user id and that appears to be how they set their permissions. So currently we give an employee permissions and then give him a password  at this time Traxia will generate a 4 digit id for that person. Don't know if any of this info will help.

**Comment by odiea** [comment]
Yes I did some basic cashier network testing with both of us at different terminals entering the same item and other items and had no problems with that setup. I could not test the Invoice issue because the current database I was using already has a single user issue. It can never find the last Invoice number used. It is always 2 Invoice numbers behind.

**Comment by marteserede** [comment]
Does the tracking of the opening/closing-amount of the cash in the register really belong to the multi location support-feature? I think that one actually goes in the direction of an integrated accounting tool and could be implemented seperately, or am I wrong? Maybe in the first step only tracking those register-amounts but with a structure (db-layout) which can later be the basis for keeping track of receivables and payables...
Which is not that difficult I think: the most basic layout would be something like this (double entry accounting of course):
accounts/ledgers (id, name, opening_balance)
transactions/entries (id, date, narration)
transaction-items/entry-items (id, transaction_id, account_id, amount)

**Comment by daN4cat** [comment]
Indeed, there are many different parts in this "multi" story.

**Comment by jekkos** [comment]
Beside the concept of the pos location, I will also introduce a store entity

* stores can have multiple stock locations
* grants are assigned on store level (so employees sign in to a pos_location, which is linked to a store, on which a grant is required)
* local admins can configure a store's employees. Also they can grant access to stock locations specific to that store
* admin users typically have access to all stores
* pos location has to be stored in the browser in localStorage (through a pair button on the config screen). Once this is done, a default store selection can be preselected on the login screen
* stores will have a separate module that can be accessed by admin users (which can then add new stores at will)
* reporting will still be available per store (so every local admin can see his / her data). system admins will probably be able to do reporting for multiple stocks at the same time

**Comment by odiea** [comment]
Looks great. A lot of new info to add. Should you break this into different modules and add them and then test them before you commit all or test as a lump sum.

**Comment by odiea** [comment]
Are you referring to multi terminals at 1 location or are you referring to Multiple store fronts as locations?

**Comment by jekkos** [comment]
well one store can have multiple pos locations. Stores will have a different geographical location each. I think I'll develop most of the functionality on a private branch first (in agreement with the client) and then release once it's approved by them.. but do intermediate deploys on my own server (to have some input first)

**Comment by odiea** [comment]
Sounds like a good plan. Are you going to fix the Invoice bug on suspend before or during this new branch being released.

**Comment by jekkos** [comment]
I already did a fix for and now it's in the repo as well. Maybe check if ok from your side now?

**Comment by jekkos** [comment]
I'm adding a new stores module to manage the multistore. In this module a user can add stores, but also add pos locations and stock locations to that store. I'm looking for a good UI idea to enable adding locations through a thickbox form. Currently I'm basing my work off the item kits editor, which allows one to add already existing items to a kit (through an autocomplete field). If anyone has some idea on how to extend that design a little so we can use it to add locations as well, then that would be helpful.

**Comment by jekkos** [comment]
I think I found a way that's nice and easy. If the entered search string in autocomplete doesn't exist, then an option 'add stock location' or 'add pos location' will be shown in that search box. If the user clicks this item, a stock location with that name is created and added to the store.

**Comment by jekkos** [comment]
Currently I'm finishing this feature up. store module gives a user the ability to create and add stock and pos locations. I'll need to do some more minor refactoring on the master here which I'll pull into my private branch afterwards.

**Comment by jekkos** [comment]
Yes currently I have the stores setup on my private server. However still some things to finish here

* stores can't be added yet through the new module
* address, telephone company name, now will be store properties (and be remove from the global config tab)
* stock locations and stores should have a many to many relationship (as sometimes companies use a warehouse from which separate stores can make orders)

**Comment by reachventures** [comment]
Is this module ready? I can test it real time on my cafes (www.kaapihuts.in). 
Currently, I have opensourcepos installed and running on different subdomains for each cafe locations I have and weekly we merge and consolidate sales reports. It is a hassle but it solves purpose. A multi-location will benefit.

**Comment by arionkosturi** [comment]
Any progress on this?

**Comment by jekkos** [comment]
Milestone changed to 3.0 or ahead. Although the private branch is mostly functional, some rework with current bootstrap branch will be needed before this can be released.

**Comment by pmontepagano** [comment]
Hi! Any updates on this?

**Comment by mckaygerhard** [comment]
ithink this can be made more easyle by usage of the db prefix but seem will be removed with #2361 .. each client/company can have own prefix...

**Comment by SteveIreland** [comment]
They can also have their own database and an instance of OSPOS for each store.  Any approach along these lines (including the use of db-prefix) takes it outside of the scope of the OSPOS.

I believe the requirements of this project meant an improvement to OSPOS to support multiple stores within a single instance of OSPOS.

**Comment by mckaygerhard** [comment]
with multidatabase.. how will be the consolidation of the data, i mean, day of cash etc? @SteveIreland ?

**Comment by mckaygerhard** [comment]
with multidatabase.. how will be the consolidation of the data, i mean, day of cash etc? @SteveIreland ?
commonly those kind of configurations are same owner with multiple stores.. want to cross the inventory and others data...

**Comment by SteveIreland** [comment]
https://stackoverflow.com/questions/674115/select-columns-across-different-databases

**Comment by mckaygerhard** [comment]
err ok ok @SteveIreland that was pretty sarcastic but yeah! 

i refers respect the ospos interface gui and the client perception.. but welll also i think that now ospos has grow up so many ...

**Comment by SteveIreland** [comment]
Sorry ... it was meant to be an answer.

**Comment by mckaygerhard** [comment]
dont worry @SteveIreland  my argument (including sarcastic that's really was funny) are that ospos will has now by example one instance with 3 stores.. but same owner of stores.. 

taking the suggestion of the cross querys across databases the queruys at code sintax will be very spagetty now! manteniance will be very hard if there no are a mecanish to made that..

ok, we can also said: "no.. each instance store are separated, if you wnat consolidate data, install and configure your own report solution and your own code to cross pass items to the others stores" bla bla.. and etc..

**Comment by JvalantThakrar** [comment]
I think it should be as simple as having a shop list module for master admin, login as shop link to different shop, and having shop id in each table to have the data restricted to the shop only. This helps in case of multiple franchise of a brand across different locations. I have done it upto shop list and direct shop login from master admin, working on adding shop id to each table and adding references to each model. Any views?

**Comment by objecttothis** [comment]
@opensourcepos/org I am personally interested in seeing this happen, but it sounds like we need to put our heads together to define the end result wanted and define the best way to achieve that with the data and code.  Our company is currently using stock_locations to define multiple store locations and in its simplest form it works, but there are limits to that usage.

While creating multiple databases and communicating between databases would require the least number of database schema changes and minimal code design changes, it also adds a lot of overhead in communicating between databases.  OSPOS was built to be a centralized inventory management and sales interface.  Separate databases decentralizes the database schema even if the databases are stored on the same physical server.

If the goal is to allow employees to be assigned or logged into a store location with default access to items in that store location to sell (sales), but still be able to optionally search inventory located at other locations (items) then the implementation isn't all that different from stock_locations.  It's simply an added layer and filters on searches and reports.  Most multi-store businesses will benefit from reports showing data for multiple selected locations as well.

As far as I can tell, there is added work in relation to permissions (who sees what), but not really a lot of added complication.  Just like @daN4cat is suggesting, it needs to be designed on paper, agreed on and implemented.  Count me in on doing some of the code, but I would prefer this be a group effort.  It sounds like @jekkos has already done a good amount of work on this.

**Comment by WebShells** [comment]
@objecttothis I can take a part of that if i wasnt assigned to complete dues reports..  @jekkos mentioned before that he already made alot of progress regarding this... There are several things to take into consideration, receipt printing assigning stock location address and specific phone number, different tables for people who are using OSPOS for restaurants, cashups module, etc...
I would like to go forward with this as well..

**Comment by SteveIreland** [comment]
My guess is also that those sites that are using locations are using them as a poor substitute for stores.

I'm trying to learn Laravel and I like to learn "hands on".  Of course, my first exercise of choice is to migrate OSPOS over to Laravel and my first subproject is to transform locations to represent either store locations or locations that belong to a particular store.  Fortunately I'm still employed but the impact of that is that I'm struggling to find time to spend on the effort.

I mention this because my approach for my "exercise" is to to create a Locations table that contains some of the information from general configuration so that each store has their own address, etc to be printed on invoices.   I'm assuming that @jekkos is doing something similar but won't know for sure until he shares the branch.

**Comment by daN4cat** [comment]
Re code OSPOS in Laravel from CI sounds a bit of a work.
I'm not sure how that could help making OSPOS multi location.
It's all down to how the data model is designed.

**Comment by WebShells** [comment]
I was thinking the same... That would be somehow a bunch of work...

**Comment by SteveIreland** [comment]
Right ... I wasn't suggesting that it would help in making OSPOS multi-location.  I was just offering my thoughts on the data model design.   Sorry.

I did say that it was an exercise for me to learn Laravel.  The work effort is irrelevant at this point.

**Comment by daN4cat** [comment]
To have an application in mind is a good way to learn a new language, framework or paradigm.

**Comment by daN4cat** [comment]
Laravel GraphQL server: https://github.com/nuwave/lighthouse

**Comment by SteveIreland** [comment]
Thanks for the insight.   It looks interesting and certainly compatible with Laravel which seems to have somewhat assimilated the RestAPI paradigm.  I'll dance with it for awhile.

**Comment by objecttothis** [comment]
I just did a precursory look into Laravel and associated Framework pieces.  I'm guessing that since Laravel is in itself a PHP framework that it would replace CI and we would be rewriting the application into the same language but different framework.  Is that correct?

Out of curiosity, are there other ways with the current implementation to implement offline mode and multi-store features?  multi-store seems doable to me, but I wouldn't know where to begin with an offline mode.  Both seem to be features people are interested in, but a complete rewrite might be a harder sell.

**Comment by jekkos** [comment]
What do you mean with offline mode? You can perfectly run ospos wo internet connection?

Multi store support practically means a datamodel and permissions change. Then adapt lots of queries to take the new fk into account.

**Comment by daN4cat** [comment]
@objecttothis what about having a database synchronisation with the remote?
You have a local installation of OSPOS on a RaspberryPi with read & write capabilities and you remote sync the database that is just a read only replica in a cloud.

**Comment by daN4cat** [comment]
You connect just to make the sync at your earliest convenience or in batches overnight.

**Comment by jekkos** [comment]
Indeed you could just run a cronjob that dumps your local database and pushes a copy once you have connection to the internet.

You could run the code locally from version control and push new versions from a git repo.

**Comment by mckaygerhard** [comment]
i already see that muti tenant are not a target amont years in opensourcepos.. adn as i can see this issue has many years as i explained in #2775 , but the most interesting response are in the comment of #1937 "i will not work for free" of @daN4cat the "lead architect" and i understand their reasons.. we all need money to buy food of course!.  So! @jekkos and others please read carefully my proposal is to use two differentiators and i tested in some years ... 

1. first: will be the prefix the prefix will have "master" or "ospos" if the prefix database are "ospos" and has same structure .. there's no a master database.. it's a deploy.. so the same instalation can filter the problem if there's no good installation made!
2. and the most important is **a "deploy comun" that will be the deploy indicator (a hash for each installation)** to bring the right multi tenant identification without mayor development on the current coding
3. now each "location" (deploy) can be configure as mysql-slave db, of course we will need a master one, easy!

Basically this behaviour let the main project opensourcepos without mayor changes.. only one column will be added to each table (the deploy column, please dont named "id_deploy".. just "deploy" right) in this way each replication of that poor mysql handler can be master-slave to a central db that will have everything this will have prox and cons:

**PROS**: each installation will have a hash id that will help in debug problems amont the deploys adn where are.. we can described each deploy linkig to it id hash, also, there all these reports... evidently this will bring the benefit of eliminating these annoying reports that the system has and so the installer will have to deploy his own reporting system, put inclusivelly we can let these reports as minimal flavors of, and if some people want more granular ones.. must setup a report system in the master database.. also all teh cash operations and accountant could be made in the master database.

**CONS**: of course the main database or master will need a powered machine, mysql in default isntallations cannot handle a production deploy.. i have deploy a powered suggested setup as i posted in the alpine mariadb wiki: https://wiki.alpinelinux.org/wiki/Production_DataBases_:_mysql#Production_settings (each of these setup depends on a system with 4G of ram.. most setting must be re-calculed to amount of ram adn disk). NOTE: of course also some tables will must be partitioned in the main db.

Linked isssues: as in #1670 #1113 #1664 #2226 and #2775 but the most interesting response are in last comments of #1937

**Comment by mouradsme** [comment]
@jekkos which branch was it iydmma?
